web/js/codeworld_shared.js

```diff
@@ -122,13 +122,111 @@ var hintBlacklist = [
     "Pi",
 ];
 
-var codeWorldBuiltinsDocs = {};
+
+// codeWorldDocs is variable containing annotations and documentation
+// of builtin and user-defined variables.
+// Expected format:
+// codeWorldDocs = {
+//   codeWorldLogo: {
+//     annotation: "Picture",
+//     doc: "The CodeWorld logo."
+//   }
+// }
+
+var codeWorldDocs = {};
+
+function renderHover(keyword) {
+    var elem = document.createElement('div')
+
+    if (!codeWorldDocs[keyword]){
+        return null;
+    };
+    elem.title = keyword;
+    var keywordData = codeWorldDocs[keyword];
+
+    // Variable name + annotaion
+    var topLine = document.createElement("div");
+    topLine.className = "hover-decl";
+    var wordElem = document.createElement("span");
+    wordElem.className = "hint-word";
+    wordElem.appendChild(document.createTextNode(keyword));
+    topLine.appendChild(wordElem);
+    if (keywordData.annotation) {
+        var annotation = document.createElement('div');
+        topLine.appendChild(document.createTextNode(" :: " + keywordData.annotation))
+    };
+    elem.appendChild(topLine)
+
+    // Docs
+    if (keywordData.doc) {
+        var docElem = document.createElement('div');
+        docElem.innerHTML = keywordData.doc;
+        docElem.className = "hover-doc";
+        elem.appendChild(docElem)
+    };
+    return elem;
+};
+
+function applyCodemirrorChange(line, change) {
+    if (change.origin === "+input"){
+        return line.slice(0, change.from.ch) + change.text.join('') + line.slice(change.from.ch);
+    } else if (change.origin === "+delete"){
+        return line.slice(0, change.from.ch) + line.slice(change.from.ch + 1);
+    } else if (["setValue", "complete"].indexOf(change.origin) != -1){
+        return;
+    } else {
+        console.log("Event \"" + change.origin + "\" not handled.");
+        return;
+    }
+}
+
+function getWordAndAnnotation(line) {
+    var annotation = null,
+        word = null;
+    // there is top-level assignment
+    if (/^\S+\s*=/.test(line)) {
+        word = line.split("=")[0].trim()
+    };
+    // there is top level type annotation
+    if (/^\S+\s*::/.test(line)) {
+        var splitted = line.split("::");
+        word = splitted[0].trim()
+        annotation = splitted[1].trim()
+    };
+    return {
+        word: word,
+        annotation: annotation
+    }
+}
+
+function updateDocs(cm, change) {
```

Comment:
I am skeptical of this whole approach.  I'd be more comfortable with something like setting a timer when the document changes, and re-parsing the whole document when the timer expires.  My fear is that this is likely to get out of sync, leaving us in a situation where the docs or autocomplete are wrong, and the only way to fix it is to reload the document.

Reply:
I suspect this is also the reason you've chosen to ignore lots of edit types above, which is more evidence it's probably not the right way to go.

Reply:
Got it. Reasonable.
I'll try to rewrite from event-oriented approach to single parsing function + timer.

Reply:
Done. Take a look please.